### PR TITLE
Groth16 verifier pallet

### PR DESCRIPTION
**Pull request by lgiussan**

### CODE REVIEW

File: pallets/settlement-groth16/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                Err(Groth16Error::VerifyError) => Err(Error::<T>::VerifyError).map_err(Into::into),
+                Err(Groth16Error::InvalidInput) => {
+                    Err(Error::<T>::InvalidInput).map_err(Into::into)
+                }
+            }
+        }
+    }
+
+    pub fn compute_groth16_hash(vk: &VerificationKeyWithCurve, input: &[Scalar]) -> H256 {
+        const PREFIX: &str = "groth16-";
```
Comment (DanieleDiBenedetto) [inline pallets/settlement-groth16/src/lib.rs:107]: Please let's use the same approach as in:
https://github.com/HorizenLabs/zkVerify/blob/main/pallets/settlement-fflonk/src/lib.rs#L170
Via submission of a constant

File: pallets/settlement-groth16/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                Ok(false) => Err(Error::<T>::VerifyError).map_err(Into::into),
+                Err(Groth16Error::InvalidProof) => {
+                    Err(Error::<T>::InvalidProof).map_err(Into::into)
+                }
+                Err(Groth16Error::InvalidVerificationKey) => {
+                    Err(Error::<T>::InvalidVerificationKey).map_err(Into::into)
+                }
+                Err(Groth16Error::VerifyError) => Err(Error::<T>::VerifyError).map_err(Into::into),
+                Err(Groth16Error::InvalidInput) => {
+                    Err(Error::<T>::InvalidInput).map_err(Into::into)
```
Comment (DanieleDiBenedetto) [inline pallets/settlement-groth16/src/lib.rs:100]: Can we make this more succinct ?

File: pallets/settlement-groth16/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                }
+            }
+        }
+    }
```
Comment (DanieleDiBenedetto) [inline pallets/settlement-groth16/src/lib.rs:104]: If it's not too much work, let's add the register_vk extrinsic as well. You can take inspiration from the fflonk pallet already merged. We will refactor everything later when Verifier generalization will be available

File: pallets/settlement-groth16/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        proof,
+        vk,
+        input: inputs,
+    }
+    .get_dispatch_info();
+
+    assert_eq!(info.pays_fee, Pays::Yes);
+    assert_eq!(
+        info.weight,
+        MockWeightInfo::submit_proof_bn254(num_inputs as u32)
```
Comment (DanieleDiBenedetto) [inline pallets/settlement-groth16/src/tests.rs:57]: Shouldn't we test also usage of proper weights for bls12_381 ?

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]
Minor comments, overall LGTM !

File: pallets/settlement-groth16/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (la10736) [inline pallets/settlement-groth16/src/lib.rs:1]: You missed the headers here

File: pallets/settlement-groth16/src/lib.rs (hunk 3 above)
Comment (la10736) [reply pallets/settlement-groth16/src/lib.rs:104]: Maybe is better to wait that the `pallet-verifiers` land

File: pallets/settlement-groth16/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    use sp_std::vec::Vec;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type OnProofVerified: OnProofVerified;
+        type WeightInfo: WeightInfo;
+        const MAX_NUM_INPUTS: usize;
```
Comment (la10736) [inline pallets/settlement-groth16/src/lib.rs:44]: You can use constant paramenter : `#[pallet::constant]`

File: pallets/settlement-groth16/src/benchmarking.rs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+use crate::Pallet as SettlementGroth16Pallet;
+use frame_benchmarking::v2::*;
+use frame_system::RawOrigin;
+
+#[benchmarks]
+mod benchmarks {
+    use super::*;
+
+    #[benchmark]
+    fn submit_proof_bn254(n: Linear<0, 16>) {
```
Comment (la10736) [inline pallets/settlement-groth16/src/benchmarking.rs:14]: Maybe here you can access to `<T as Config>::MAX_NUM_INPUTS`

File: pallets/settlement-groth16/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,114 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
```
Comment (la10736) [inline pallets/settlement-groth16/src/lib.rs:2]: Add also

```
#![deny(missing_docs)]
```

File: pallets/settlement-groth16/src/tests.rs (hunk 4 above)
Comment (la10736) [reply pallets/settlement-groth16/src/tests.rs:57]: +1

File: pallets/settlement-groth16/src/tests.rs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        let (_, vk, inputs) = Groth16::get_instance(10, Some(42), Curve::Bn254);
+
+        assert!(
+            SettlementGroth16Pallet::submit_proof(RuntimeOrigin::signed(1), proof, vk, inputs,)
+                .is_err()
+        );
+
+        let events = mock::System::events();
+        assert_eq!(events.len(), 0);
+    });
```
Comment (la10736) [inline pallets/settlement-groth16/src/tests.rs:39]: You can use `assert_noop!()` here (look settlement-fflonk as example) to assert that you have an error and the state is unchanged

File: pallets/settlement-groth16/src/tests.rs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+use crate::weight::WeightInfo;
+use crate::{compute_groth16_hash, groth16::Curve, groth16::Groth16};
+use frame_support::dispatch::{GetDispatchInfo, Pays};
+
+#[test]
+fn valid_proof_passes_verification_and_is_notified() {
+    new_test_ext().execute_with(|| {
+        let (proof, vk, inputs) = Groth16::get_instance(10, None, Curve::Bls12_381);
+        let hash = compute_groth16_hash(&vk, &inputs);
+        assert!(
```
Comment (la10736) [inline pallets/settlement-groth16/src/tests.rs:12]: You can use `assert_ok` here

File: pallets/settlement-groth16/src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,59 @@
+use crate::mock;
+use crate::mock::*;
```
Comment (la10736) [inline pallets/settlement-groth16/src/tests.rs:2]: just the second one is enought

Comment (la10736) [reply pallets/settlement-groth16/src/tests.rs:2]:
```suggestion
use crate::mock::*;
use crate::weight::WeightInfo;
```

File: pallets/settlement-groth16/src/tests.rs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            SettlementGroth16Pallet::submit_proof(RuntimeOrigin::signed(1), proof, vk, inputs)
+                .is_ok()
+        );
+
+        let events = mock::System::events();
+        assert_eq!(events.len(), 1);
+
+        mock::System::assert_last_event(
+            crate::mock::on_proof_verified::pallet::Event::NewProof { value: hash }.into(),
+        );
```
Comment (la10736) [inline pallets/settlement-groth16/src/tests.rs:22]:
```suggestion
        assert_eq!(System::events().len(), 1);

        System::assert_last_event(
            on_proof_verified::pallet::Event::NewProof { value: hash }.into(),
        );
```

**Review by la10736 — CHANGES_REQUESTED** [review]
Just little changes

File: pallets/settlement-groth16/src/lib.rs (hunk 6 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/lib.rs]: Fixed in https://github.com/HorizenLabs/zkVerify/pull/66/commits/222bb7568e108e7fe40ebbb95fcbdd835926f01d

File: pallets/settlement-groth16/src/benchmarking.rs (hunk 7 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/benchmarking.rs]: Fixed in https://github.com/HorizenLabs/zkVerify/pull/66/commits/222bb7568e108e7fe40ebbb95fcbdd835926f01d

File: pallets/settlement-groth16/src/lib.rs (hunk 2 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/lib.rs]: Addressed in https://github.com/HorizenLabs/zkVerify/pull/66/commits/6935dc9424b0630ec8b3ef2b3051ebc52c31fd22

File: pallets/settlement-groth16/src/lib.rs (hunk 5 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/lib.rs:16]: Added them in https://github.com/HorizenLabs/zkVerify/pull/66/commits/49dfe05d30aee834f76b2af36511bad3e4158d1b

File: pallets/settlement-groth16/src/lib.rs (hunk 8 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/lib.rs:17]: Added some documentation comments in https://github.com/HorizenLabs/zkVerify/pull/66/commits/0869642be957d026f8a90bb5086d230859d2a69c  :-P

File: pallets/settlement-groth16/src/tests.rs (hunk 4 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/tests.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/484bf5bc5313c1a7f529be45c20272069b00249e

File: pallets/settlement-groth16/src/tests.rs (hunk 9 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/tests.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/484bf5bc5313c1a7f529be45c20272069b00249e

File: pallets/settlement-groth16/src/tests.rs (hunk 10 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/tests.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/484bf5bc5313c1a7f529be45c20272069b00249e

File: pallets/settlement-groth16/src/tests.rs (hunk 11 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/tests.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/484bf5bc5313c1a7f529be45c20272069b00249e

File: pallets/settlement-groth16/src/tests.rs (hunk 12 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/tests.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/484bf5bc5313c1a7f529be45c20272069b00249e

File: pallets/settlement-groth16/src/groth16_generic.rs [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                .err()
+                .unwrap(),
+            Groth16Error::VerifyError
+        )
+    }
+
+    #[test]
+    fn verify_succeeds_bn254() {
+        verify_succeeds::<Bn254>()
+    }
```
[the remark this answers is not in the document]
Comment (lgiussan) [reply pallets/settlement-groth16/src/groth16_generic.rs]: Neat! Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/484bf5bc5313c1a7f529be45c20272069b00249e

**Comment by la10736** [comment]
I noted that there is any  E2E test at all

File: pallets/settlement-groth16/src/lib.rs (hunk 1 above)
Comment (lgiussan) [reply pallets/settlement-groth16/src/lib.rs]: Changed the definition of the hash in https://github.com/HorizenLabs/zkVerify/pull/66/commits/984e7469ae155e744aebfaff57e4a897e48ba9ac according to the new convention

**Comment by DanieleDiBenedetto** [comment]
> I noted that there is any E2E test at all

Yep. Should be enough to add a couple of lines [here](https://github.com/HorizenLabs/zkVerify/blob/main/e2e-tests/js_scripts/0005-proofPath_rpc.js)

File: runtime/src/tests.rs [hunk 14]
```diff
@@ -274,6 +303,48 @@ mod use_correct_weights {
...
+
+        const MAX_NUM_INPUTS: u32 = <Runtime as pallet_settlement_groth16::Config>::MAX_NUM_INPUTS;
+        assert_eq!(
+            <Runtime as pallet_settlement_groth16::Config>::WeightInfo::submit_proof_bls12_381(
+                MAX_NUM_INPUTS
+            ),
+            crate::weights::pallet_settlement_groth16::NHWeight::<Runtime>::submit_proof_bls12_381(
+                MAX_NUM_INPUTS
+            )
+        );
```
Comment (la10736) [inline runtime/src/tests.rs:345]: You should just test the you'd linked the correct implementation and not the implementation itself: just the first assert of the first test is enought

File: pallets/settlement-groth16/src/data_structures.rs [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+                <E::G1 as UniformRand>::rand(&mut rng).into(),
+                <E::G1 as UniformRand>::rand(&mut rng).into(),
+            ],
+        };
+
+        let serialized_vk: VerificationKey = vk.clone().try_into().unwrap();
+        let deserialized_vk: ark_groth16::VerifyingKey<E> = serialized_vk.try_into().unwrap();
+
+        assert_eq!(vk, deserialized_vk);
+    }
```
[the remark this answers is not in the document]
Comment (lgiussan) [reply pallets/settlement-groth16/src/data_structures.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/bbe356859ea2c1000cd8e204f22ec03b025e1e8b (also useful for e2e test)

File: runtime/src/tests.rs (hunk 14 above)
Comment (lgiussan) [reply runtime/src/tests.rs]: Done in https://github.com/HorizenLabs/zkVerify/pull/66/commits/978593af40cb6c4ce8401657448b97cb8929c29c

**Comment by lgiussan** [comment]
> I noted that there is any E2E test at all

Added e2e test in https://github.com/HorizenLabs/zkVerify/pull/66/commits/8fa1c541c8dfc23ea47468e9eb93ff026c93b900

**Review by DanieleDiBenedetto — APPROVED** [review]

**Review by la10736 — APPROVED** [review]

**Pull request merged**